src/react-redux/provider.js

```diff
@@ -1,42 +1,136 @@
 import { Component } from 'react';
 import { connect } from 'react-redux';
+import debugFactory from 'debug';
 import PropTypes from 'prop-types';
 import { get } from 'lodash';
+import * as actions from './actions';
 
-class FreshDataReduxProvider extends Component {
+const debug = debugFactory( 'fresh-data:provider' );
+
+export class FreshDataReduxProvider extends Component {
 	static propTypes = {
 		children: PropTypes.node.isRequired,
+		apis: PropTypes.object.isRequired,
 		rootPath: PropTypes.oneOfType( [
 			PropTypes.arrayOf( PropTypes.oneOfType( [ PropTypes.string, PropTypes.number ] ) ),
 			PropTypes.string,
 		] ),
-		update: PropTypes.func,
+		rootData: PropTypes.object.isRequired,
+		dataRequested: PropTypes.func.isRequired,
+		dataReceived: PropTypes.func.isRequired,
+		errorReceived: PropTypes.func.isRequired,
 	};
 
 	static defaultProps = {
 		rootPath: [ 'freshData' ],
-		update: () => {}, // TODO: Hook in to registry module by default.
 	};
 
+	static childContextTypes = {
+		getApiClient: PropTypes.func.isRequired,
+	};
+
+	constructor( props ) {
+		super( ...arguments );
+		this.apisByName = new Map();
+		this.namesByApi = new Map();
+		this.updateNeeded = false;
+		this.update( props );
+	}
+
+	getChildContext() {
+		return { getApiClient: this.getApiClient };
+	}
+
 	componentDidMount() {
-		const { rootData, update } = this.props;
-		update( rootData );
+		this.update( this.props );
 	}
 
 	componentDidUpdate() {
-		const { rootData, update } = this.props;
-		update( rootData );
+		this.update( this.props );
+	}
+
+	shouldComponentUpdate() {
+		return this.updateNeeded;
+	}
+
+	update( props ) {
+		const { apis, rootData } = props;
+
+		if ( this.lastApis !== apis ) {
```

Comment:
Since `this.lastApis !== apis` is used here and in the next conditional should we assign the response to a `const`?